### PR TITLE
add support for dynamic dependencies with setuptools

### LINKED ISSUE #421: KeyError when pyproject.toml dependencies are dynamically read from requirements.txt
**Describe the bug**

I usually have the following structure in my projects:

```toml
# pyproject.toml
[build-system]
requires = ["setuptools>=67.6.0"]
build-backend = "setuptools.build_meta"
[project]
dynamic = ["dependencies"]
# no `dependencies` key
[tool.setuptools.dynamic]
dependencies = { file = ["requirements.txt"] }
```

And then the projects requirements are listed in the `requirements.txt` file. This is useful because IDEs will usually syntax-highlight PEP 440 syntax in requirements files, but not in strings inside TOML tables.

When I run `deptry .` with such a project, it crashes with the following traceback:

```
Traceback (most recent call last):
  File ".../bin/deptry", line 8, in <module>
    sys.exit(deptry())
  File ".../python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
[...]
  File ".../python3.10/site-packages/deptry/core.py", line 63, in run
    dependencies_extract = self._get_dependencies(dependency_management_format)
  File ".../python3.10/site-packages/deptry/core.py", line 142, in _get_dependencies
    return PEP621DependencyGetter(self.config, self.package_module_name_map).get()
  File ".../python3.10/site-packages/deptry/dependency_getter/pep_621.py", line 41, in get
    dependencies = [*self._get_dependencies(), *itertools.chain(*self._get_optional_dependencies().values())]
  File ".../python3.10/site-packages/deptry/dependency_getter/pep_621.py", line 48, in _get_dependencies
    dependency_strings: list[str] = pyproject_data["project"]["dependencies"]
KeyError: 'dependencies'
```

**To Reproduce**

Using a pyproject.toml file as described above (with no `dependencies` key inside the top-level `project` table, but mentioning `"dependencies"` as a `dynamic` metadata), inside a Python 3.10 virtual environment, run:

```
pip install -e .
pip install deptry
deptry .
```

**Expected behavior**

Either:
- Support for the the `tool.setuptools.dynamic.dependencies.file` idiom (see https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata); or
- A clear error message saying that this idiom is not supported.

**System:**

-   OS: Ubuntu 20.04 inside WSL (Windows 10)
-   Language Version: Python 3.10.12
-   Poetry version: [not using poetry or any other packaging framework]

### PR DESCRIPTION
**Pull request by lispandfound**
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

This pull request adds support for dependencies dynamically specified when using setuptools. This fixes #421. The changes should not affect users of setuptools that do not use dynamic dependencies, as in this case it will revert to the PEP621 compliant dependency checker.

Known issue: if the specified file to source dependencies from does not exist, this code will crash. @fpgmaas what is the best way to present this error to the user?

### CODE REVIEW

**Comment by lispandfound** [comment]
Will add tests and things later.

**Comment by fpgmaas** [comment]
Hey! Thanks for your contribution. I will check it out soon, I first need to sort out the bug in the CI/CD pipeline :)

**Comment by fpgmaas** [comment]
Could you please add a functional test? i.e. 
- Add a directory with an example project structure in [`tests/data'](https://github.com/fpgmaas/deptry/tree/main/tests/data)
- Add a corresponding test under [`tests/functional/cli'](https://github.com/fpgmaas/deptry/tree/main/tests/functional/cli)

**Comment by fpgmaas** [comment]
Looks like there were a few formatting issues; https://github.com/fpgmaas/deptry/actions/runs/9409803890/job/25970899533?pr=724. You can resolve this by installing the pre-commit hooks locally with `poetry run pre-commit install`.

**Comment by fpgmaas** [comment]
> Known issue: if the specified file to source dependencies from does not exist, this code will crash. @fpgmaas what is the best way to present this error to the user?

I think in this case, it would be best to catch the `FileNotFoundError`, display a logging message to the user and exit.

File: docs/usage.md [hunk 1]
```diff
@@ -37,10 +37,16 @@ To determine the project's dependencies, _deptry_ will scan the directory it is
 2. If a `pyproject.toml` file with a `[tool.pdm.dev-dependencies]` section is found, _deptry_ will assume it uses PDM and extract:
     - dependencies from `[project.dependencies]` and `[project.optional-dependencies]` sections
     - development dependencies from `[tool.pdm.dev-dependencies]` section and from the groups under `[project.optional-dependencies]` passed via the [`--pep621-dev-dependency-groups`](#pep-621-dev-dependency-groups) argument.
-3. If a `pyproject.toml` file with a `[project]` section is found, _deptry_ will assume it uses [PEP 621](https://peps.python.org/pep-0621/) for dependency specification and extract:
+3. If a `pyproject.toml` file containing the following is found:
+    1. a `[build-system]` section containing `build-backend = "setuptools.build_meta"`,
+    2. a `[project]` section a list of `dynamic` values containing `dependencies`,
+    3. a `[tools.setuptools.dynamic]` section containing `dependencies = { file = some_requirements_file }`,
+   then _deptry_ will assume it uses setuptools project with dynamic dependencies and extract dependencies
```
Comment (fpgmaas) [inline docs/usage.md:44]: this does not render as expected in mkdocs. It shows as a continuation of point 3.

File: python/deptry/dependency_getter/builder.py [hunk 2]
```diff
@@ -45,21 +45,89 @@ def build(self) -> DependencyGetter:
...
-                self.config, self.package_module_name_map, requirements_files, self.requirements_files_dev
+                self.config,
+                self.package_module_name_map,
+                requirements_files,
+                self.requirements_files_dev,
             )
 
         raise DependencySpecificationNotFoundError(self.requirements_files)
 
+    def _project_uses_setuptools(self, pyproject_toml: dict[str, Any]) -> bool:
```
Comment (mkniewallner) [inline python/deptry/dependency_getter/builder.py:80]:
```suggestion
    @staticmethod
    def _project_uses_setuptools(pyproject_toml: dict[str, Any]) -> bool:
```

File: python/deptry/dependency_getter/builder.py [hunk 3]
```diff
@@ -45,21 +45,89 @@ def build(self) -> DependencyGetter:
...
+                )
+                return False
+        except KeyError:
+            logging.debug(
+                "pyproject.toml does not contain a build-system.build-backend entry, "
+                "so setuptools is not used to specify the project's dependencies."
+            )
+            return False
+
+    def _project_uses_dynamic_dependencies(self, pyproject_toml: dict[str, Any]) -> bool:
```
Comment (mkniewallner) [inline python/deptry/dependency_getter/builder.py:103]:
```suggestion
    @staticmethod
    def _project_uses_dynamic_dependencies(pyproject_toml: dict[str, Any]) -> bool:
```

File: python/deptry/dependency_getter/builder.py [hunk 4]
```diff
@@ -45,21 +45,89 @@ def build(self) -> DependencyGetter:
...
+                return True
+        except KeyError:
+            logging.debug(
+                "pyproject.toml either does not contain a project.dynamic entry or "
+                "tool.setuptools.dynamic.dependencies.file entry, so dynamic "
+                "dependencies are not used."
+            )
+            return False
+
+    def _project_dynamic_requirements_file(self, pyproject_toml: dict[str, Any]) -> tuple[str, ...]:
```
Comment (mkniewallner) [inline python/deptry/dependency_getter/builder.py:128]:
```suggestion
    @staticmethod
    def _project_dynamic_requirements_file(pyproject_toml: dict[str, Any]) -> tuple[str, ...]:
```

File: python/deptry/dependency_getter/builder.py (hunk 2 above)
Comment (mkniewallner) [inline python/deptry/dependency_getter/builder.py:80]: Could you move the 3 new methods between `_project_uses_pdm` and `_project_uses_pep_621`? That way we use the same order as the order methods are called from `build` method.

File: python/deptry/dependency_getter/builder.py (hunk 4 above)
Comment (mkniewallner) [inline python/deptry/dependency_getter/builder.py:128]: Maybe we can remove this method and make `_project_uses_dynamic_dependencies` return a `tuple[bool, tuple[str, ...]]`, like https://github.com/fpgmaas/deptry/blob/5756281f54a1e3395300b78251a50b9563c770fd/python/deptry/dependency_getter/builder.py#L119?

File: docs/usage.md [hunk 5]
```diff
@@ -37,10 +37,18 @@ To determine the project's dependencies, _deptry_ will scan the directory it is
...
     - dependencies from `[project.dependencies]` and `[project.optional-dependencies]` sections
     - development dependencies from `[tool.pdm.dev-dependencies]` section and from the groups under `[project.optional-dependencies]` passed via the [`--pep621-dev-dependency-groups`](#pep-621-dev-dependency-groups) argument.
-3. If a `pyproject.toml` file with a `[project]` section is found, _deptry_ will assume it uses [PEP 621](https://peps.python.org/pep-0621/) for dependency specification and extract:
+3. If a `pyproject.toml` file containing the following is found:
+    1. a `[build-system]` section containing `build-backend = "setuptools.build_meta"` and,
+    2. a `[project]` section with a `dynamic` key, where the corresponding list includes `"dependencies"` and,
+    3. a `[tool.setuptools.dynamic]` section containing `dependencies = { file = [some_requirements_file] }`,
+
+    then _deptry_ will assume the project uses setuptools with dynamic dependencies. It will extract dependencies 
+    from the file specified in the `[tool.setuptools.dynamic]` section.
```
Comment (mkniewallner) [inline docs/usage.md:46]:
```suggestion
    from the files specified in the `[tool.setuptools.dynamic]` section.
```
since this could be a list of files?

**Pull request closed without merging**